### PR TITLE
API: Switch site frame nonce to user+site value

### PR DESCRIPTION
**Pull request by mmtr**
#### Changes proposed in this Pull Request:

On D27129-code we're changing the way the frame nonces are generated to use both the user ID and site ID. 

Since that will be an breaking change for the API (p7jreA-2hw-p2), we need to version the `/sites/:site` endpoint (which is where the frame nonces are serialized) in order to keep using the nonce generated from only the site ID on the existing versions (1.0 and 1.2) and switch to the user+site value on the new version (1.3).

Given Fusion is not able to apply these changes automatically, this will also need a diff in WP.com: D27153-code

#### Testing instructions:

- Make an authenticated call to `https://public-api.wordpress.com/rest/v1.2/me/sites?site_visibility=all&include_domain_only=true&site_activity=active&fields=ID,URL,options&options=frame_nonce_site_only,frame_nonce`.
- Verify that both `frame_nonce` and `frame_nonce_site_only` are returned with the same value for all WP.com and JP sites.
- Make an authenticated call to `https://public-api.wordpress.com/rest/v1.3/me/sites?site_visibility=all&include_domain_only=true&site_activity=active&fields=ID,URL,options&options=frame_nonce_site_only,frame_nonce`.
- For JP sites, check that `frame_nonce` has a different value now but `frame_nonce_site_only` still have the same value as before. For WP.com sites, both `frame_nonce` and `frame_nonce_site_only` still have the same value as before.

Repeat steps but replacing the API calls from `/me/sites` to `/sites/<SITE_SLUG_OR_ID>`.

#### Proposed changelog entry for your changes:

WPCOM API: Switch site frame nonces to new user and site specific values.

### CODE REVIEW

**Comment by mdawaffe** [comment]
> nonces are generated to use both the user ID and site ID.

 Why is this a breaking change in the API? Has Jetpack ever used the previous form of frame nonces?

**Comment by mmtr** [comment]
> > nonces are generated to use both the user ID and site ID.
> 
> Why is this a breaking change in the API? Has Jetpack ever used the previous form of frame nonces?

It will affect the WordPress.com desktop and mobile apps as they use the _old site-specific nonces_ for previewing posts.

Actually, we already deployed this change without versioning the API and had to revert it when we realized that the post previews were not working on those apps (p7jreA-2hw-p2).

File: json-endpoints/class.wpcom-json-api-get-site-v1-3-endpoint.php [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+		'query_parameters'   => array(
+			'context' => false,
+		),
+		'response_format'    => WPCOM_JSON_API_GET_Site_V1_3_Endpoint::$site_format,
+		'example_request'    => 'https://public-api.wordpress.com/rest/v1.3/sites/en.blog.wordpress.com/',
+	)
+);
+
+class WPCOM_JSON_API_GET_Site_V1_3_Endpoint extends WPCOM_JSON_API_GET_Site_V1_2_Endpoint {
+	protected function render_option_keys( &$options_response_keys ) {
```
Comment (kwight) [inline json-endpoints/class.wpcom-json-api-get-site-v1-3-endpoint.php]: Why is this needed by reference?

Comment (mmtr) [reply json-endpoints/class.wpcom-json-api-get-site-v1-3-endpoint.php]: I saw it is also passed by reference on the parent so I thought we need the same interface since we're extending the method?

https://github.com/Automattic/jetpack/blob/1531b954ebd392ca2693e28549b2333fd39b6a47/json-endpoints/class.wpcom-json-api-get-site-endpoint.php#L378

Comment (kwight) [reply json-endpoints/class.wpcom-json-api-get-site-v1-3-endpoint.php]: You're right, ignore me – it also throws a warning (saying as much) if you try to not use it 🙃

**Review by kwight — DISMISSED** [review]

File: sal/class.json-api-site-jetpack.php [hunk 2]
```diff
@@ -108,6 +108,10 @@ function get_frame_nonce() {
 		return false;
 	}
 
+	function get_frame_nonce_site_only() {
+		return false;
+	}
+
```
Comment (dmsnell) [inline sal/class.json-api-site-jetpack.php]: I know that none of the other functions are documented, but could we add a short explanatory comment here on why it always returns `false`? having one above in `get_frame_nonce()` would also be helpful and aid the casual passer-by as to why these two are different. I'm going to toss an example out to illustrate my point but I assume that everything I say in my example is wrong.

```php
function has_wordads() {
	return Jetpack::is_module_active( 'wordads' );
}

/**
 * Prevents creating an iframe nonce for connected sites (I have no idea what this is here for)
 *
 * **Note**: On WordPress.com there is no nonce because tigers
 */
function get_frame_nonce() {
	return false;
}

/**
 * Prevents creating a site-only iframe nonce for connected sites
 *
 * The site-only nonce is used in the iframe'd block editor
 * and exists because of user authentication (please write something meaningful here instead)
 * @see Some Reference To Documentation On This
 *
 * **Note**: On WordPress.com there is no nonce because bears
 */
function get_frame_nonce_site_only() {
	return false;
}
```

File: sal/class.json-api-site-jetpack.php [hunk 3]
```diff
@@ -104,10 +104,22 @@ function has_wordads() {
 		return Jetpack::is_module_active( 'wordads' );
 	}
 
+	/**
+	 * Creates a nonce to be used with iframed block editor requests to a Jetpack site.
+	 * Disabled on the Jetpack side.
```
Comment (dmsnell) [inline sal/class.json-api-site-jetpack.php]: sorry to be an obstacle here, but _why_ is it disabled and is _this_ the Jetpack side?

maybe something like this…

```php
/*
 * Creates a nonce to allow an iframed block editor to make authenticated requests to a Jetpack site.
 *
 * Disabled here in Jetpack because the nonce is only needed when making calls from the dotcom domains,
 * not when making calls on the same origin as we would be doing from `wp-admin/post.php` and friends.
 */
```

**Review by kwight — DISMISSED** [review]

**Comment by kwight** [comment]
@jeherve Would you or someone else mind a final look at this? This PR will bring JP back into sync with recent work on the WP.com side. 

- D26952-code
- D26953-code
- D26956-code
- D27153-code

**Comment by obenland** [comment]
@Automattic/jetpack-crew When one of you gets a chance, would you mind ✅ this PR to help unblock us from posting a Call for Testing?

**Review by jeherve — APPROVED** [review]
This seems to work well in my tests. Merge when ready 👍
